ellama.el

```diff
@@ -729,48 +730,171 @@ If EPHEMERAL non nil new session will not be associated with any file."
+  (ignore mode)
+  (with-slots (name) element
+    (format "[[%s][%s]]"
+	    (replace-regexp-in-string
+	     "(\\(.?*\\)) \\(.*\\)" "info:\\1#\\2" name)
+	    (if (and ellama-chat-translation-enabled
+		     (not ellama--current-session))
+		(ellama--translate-string name)
+	      name))))
+
+;; Text context element
+
+(defclass ellama-context-element-text (ellama-context-element)
+  ((content :initarg :content :type string))
+  "A structure for holding information about a context element.")
+
+(cl-defmethod ellama-context-element-extract
+  ((element ellama-context-element-text))
+  "Extract the content of the context ELEMENT."
+  (oref element content))
+
+(cl-defmethod ellama-context-element-format
+  ((element ellama-context-element-text) (mode (eql 'markdown-mode)))
+  "Format the context ELEMENT for the major MODE."
+  (ignore mode)
+  (oref element content))
+
+(cl-defmethod ellama-context-element-format
+  ((element ellama-context-element-text) (mode (eql 'org-mode)))
+  "Format the context ELEMENT for the major MODE."
+  (ignore mode)
+  (oref element content))
+
 ;;;###autoload
 (defun ellama-context-add-file ()
   "Add file to context."
   (interactive)
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session))
-	    (file-name (read-file-name "Select file: " nil nil t)))
-      (push (cons 'file file-name) (ellama-session-context session))
-    (push (cons 'file file-name) ellama--new-session-context)))
+  (let* ((file-name (read-file-name "Select file: " nil nil t))
+         (element (ellama-context-element-file :name file-name)))
+    (ellama-context-element-add element)))
 
 ;;;###autoload
 (defun ellama-context-add-buffer (buf)
   "Add BUF to context."
   (interactive "bSelect buffer: ")
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session)))
-      (push (cons 'buffer buf) (ellama-session-context session))
-    (push (cons 'buffer buf) ellama--new-session-context)))
+  (let ((element (ellama-context-element-buffer :name buf)))
+    (ellama-context-element-add element)))
 
 ;;;###autoload
 (defun ellama-context-add-selection ()
   "Add file to context."
   (interactive)
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session))
-	    ((region-active-p))
-	    (content (buffer-substring-no-properties (region-beginning) (region-end))))
-      (push (cons 'text content) (ellama-session-context session))
-    (push (cons 'text content) ellama--new-session-context)))
+  (if (region-active-p)
+      (let* ((content (buffer-substring-no-properties (region-beginning) (region-end)))
+             (element (ellama-context-element-text :content content)))
+        (ellama-context-element-add element))
+    (user-error "No active region")))
```

Comment:
```suggestion
    (warn "No active region")))
```

Reply:
Why `warn` and not `user-error` though? Until now I thought `user-error` is the way to display a message to the user if they did something wrong, since it does not open the debugger and it displays a message in the echo error. I would also argue it is a user error, since they user did not mark a region.

I just learned through your snippet of the `warn` function (or forgot about it) and tried evaluating `(warn "Hello")`. It returned `t` and I was surprised to see no message. I had to read the documentation about `warn` to realize the message shows up in the `*Warnings*` buffer. I have actually configured the `warning-minimum-level` to be `:emergency`. I remember doing this because I got many native compilation warning in the early days of that feature. 

So maybe I should really lower that level, but I'm wondering how other people have this configured. And if they do something like I did, they don't see this message, and might get confused. 
 
Wdyt?